File: src/apps/kernel/arch/host/Platform.cc
/*
 * Copyright (C) 2015, Nils Asmussen <nils@os.inf.tu-dresden.de>
 * Economic rights: Technische Universitaet Dresden (Germany)
 *
 * This file is part of M3 (Microkernel-based SysteM for Heterogeneous Manycores).
 *
 * M3 is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License version 2 as
 * published by the Free Software Foundation.
 *
 * M3 is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
 * General Public License version 2 for more details.
 */

#include <base/Config.h>
#include <base/Init.h>

#include <sys/mman.h>

#include "mem/MainMemory.h"
#include "DTU.h"
#include "Platform.h"

namespace kernel {

INIT_PRIO_USER(2) Platform::KEnv Platform::_kenv;

Platform::KEnv::KEnv() {
    // no modules
    mods[0] = 0;

    // init PEs
    pe_count = PE_COUNT;
    for(int i = 0; i < PE_COUNT; ++i)
        pes[i] = m3::PEDesc(m3::PEType::COMP_IMEM, m3::PEISA::X86, 1024 * 1024);

    const size_t TOTAL_MEM   = 512 * 1024 * 1024;

    // create memory
    uintptr_t base = reinterpret_cast<uintptr_t>(
        mmap(0, TOTAL_MEM, PROT_READ | PROT_WRITE, MAP_ANON | MAP_PRIVATE, -1, 0));

    MainMemory &mem = MainMemory::get();
    mem.add(new MemoryModule(false, 0, base, FS_MAX_SIZE));
    mem.add(new MemoryModule(true, 0, base + FS_MAX_SIZE, TOTAL_MEM - FS_MAX_SIZE));
}

peid_t Platform::kernel_pe() {
    return 0;
}
peid_t Platform::first_pe() {
    return 1;
}
peid_t Platform::last_pe() {
    return _kenv.pe_count - 1;
}

goff_t Platform::def_recvbuf(peid_t) {
    // unused
    return 0;
}

}

File: src/apps/rust/kernel/src/arch/host/platform.rs
use base::cfg;
use base::dtu::PEId;
use base::goff;
use base::kif::{PEDesc, PEISA, PEType};
use base::libc;
use core::intrinsics;
use core::ptr;

use mem;
use platform;

pub fn init() -> platform::KEnv {
    let mut kenv: platform::KEnv = unsafe { intrinsics::uninit() };

    // no modules
    kenv.mods[0] = 0;

    // init PEs
    kenv.pe_count = cfg::PE_COUNT as u64;
    for i in 0..cfg::PE_COUNT {
        kenv.pes[i] = PEDesc::new(PEType::COMP_IMEM, PEISA::X86, 1024 * 1024).value();
    }

    // create memory
    let base = unsafe {
        libc::mmap(
            ptr::null_mut(),
            cfg::MEM_SIZE,
            libc::PROT_READ | libc::PROT_WRITE,
            libc::MAP_ANON | libc::MAP_PRIVATE,
            -1,
            0
        )
    };
    assert!(base != libc::MAP_FAILED);

    let mut m = mem::MemMod::new(kernel_pe(), base as goff, cfg::MEM_SIZE);
    // allocate FS image
    m.allocate(cfg::FS_MAX_SIZE, 1).expect("Not enough space for FS image");
    mem::get().add(m);

    kenv
}

pub fn kernel_pe() -> PEId {
    0
}
pub fn first_user_pe() -> PEId {
    kernel_pe() + 1
}
pub fn last_user_pe() -> PEId {
    platform::pe_count() - 1
}

pub fn default_rcvbuf(_pe: PEId) -> goff {
    0
}
pub fn rcvbufs_size(_pe: PEId) -> usize {
    usize::max_value()
}
